### PR TITLE
rgw_sal_motr : [CORTX-32361] Fix invalid version-id-marker issue in a List Object Versions API

### PR DESCRIPTION
**Pull request by priyanka25081999**
Problem Statement:
In List Object Versions API : 
- For invalid version-id-marker value, it is showing all object versions.

Solution:
- Return `InvalidArgument` error in case of invalid version-id-marker. 
- Handle valid null version object entry in list response if user passes key-marker and version-id-marker parameter in ListObjectVersions request.

Signed-off-by: Priyanka Salunke <priyanka.s.salunke@seagate.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sjain09** [comment]
@priyanka25081999 There is one more PR #348 which is making change in the same api. can you please sync up with @vastradparayya on this

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -1300,6 +1300,24 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
     }
   }
 
+  // Return an error in case of invalid version-id-marker
+  int ret_code;
+  bufferlist bl;
+  std::string marker_key;
+  ceph::real_time marker_mtime;
+
+  marker_key = params.marker.name + '\a' + params.marker.instance;
```
Comment (sachitanands) [inline src/rgw/rgw_sal_motr.cc:1309]: pls confirm, if we need this line ? 
should it be in else block e.g. append params.marker.instance value only if its not null.

Comment (priyanka25081999) [reply src/rgw/rgw_sal_motr.cc]: done

**Comment by sjain09** [comment]
@priyanka25081999 , let's wait for PR #348 to merge. then rebase this PR and run the sanity and all ceph-s3 tests related to list bucket (list-objects).

**Review by sachitanands — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -1300,6 +1300,25 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
     }
   }
 
+  // Return an error in case of invalid version-id-marker
```
Comment (priyanka25081999) [inline src/rgw/rgw_sal_motr.cc:1303]: This code change will handle "invalid version-id-marker" issue.

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -1333,8 +1352,12 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
               ((!null_ent.key.empty() && params.marker.instance == "null" &&
                  null_ent.meta.mtime < ent.meta.mtime) ||
                (key.instance != "" &&
-                key.instance < params.marker.instance)))
-            continue;
+                key.instance < params.marker.instance))) {
```
Comment (priyanka25081999) [inline src/rgw/rgw_sal_motr.cc:1355]: This code changes will handle "valid null version-id-marker" issue.

**Comment by priyanka25081999** [comment]
In this PR, we are handling 2 issues:
1. Invalid version-id-marker fix 
2. Valid version-id-marker fix including null object versions 
    2.1 For this, similar changes has been done as a part of this PR (https://github.com/Seagate/cortx-rgw/pull/348). Hence, if we are planning to merge this #348 first, then we need to remove this changes from this current PR.

@cdeshmukh @sachitanands @sjain09 @vastradparayya @shriya-deshmukh

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -1300,6 +1300,25 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
     }
   }
 
+  // Return an error in case of invalid version-id-marker
+  int ret_code;
```
Comment (cdeshmukh) [inline src/rgw/rgw_sal_motr.cc:1304]: why cant we use rc var itself, and return same @ line 1318

**Review by cdeshmukh — APPROVED** [review]

**Review by shriya-deshmukh — APPROVED** [review]

**Comment by priyanka25081999** [comment]
Tested the deployment with below images,
- 1N k8s Deployment : https://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/Generic/job/cortx-rgw/90/ (Successful)
- Sanity (ceph s3 test suite) : https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/S3Server/job/S3-Compatibility-Test-Ceph-RGW/168/ (commit ID - 33afb4eb88c5c91caaee6bc178f44c45c052c0e8)
- Test logs : https://jts.seagate.com/secure/attachment/529146/CORTX-32361%20-%20testing%20logs.txt

**Pull request merged**